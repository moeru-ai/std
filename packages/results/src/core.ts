export interface Err<E> {
  __type__: 'err'
  error: E
}

export interface None {
  __type__: 'none'
}

export interface Ok<T> {
  __type__: 'ok'
  value: T
}

export type Option<T> = None | Some<T>

export type Result<T, E> = Err<E> | Ok<T>

export interface Some<T> {
  __type__: 'some'
  value: T
}

export const err = <T, E>(error: E): Result<T, E> => ({
  __type__: 'err',
  error,
})

export const none: Option<never> = {
  __type__: 'none',
}

export const ok = <T, E>(value: T): Result<T, E> => ({
  __type__: 'ok',
  value,
})

export const some = <T>(value: T): Option<T> => ({
  __type__: 'some',
  value,
})
