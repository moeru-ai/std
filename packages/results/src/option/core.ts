export interface None {
  __type__: 'none'
}

export type Option<T> = None | Some<T>

export interface Some<T> {
  __type__: 'some'
  value: T
}

export const none: Option<never> = {
  __type__: 'none',
}

export const some = <T>(value: T): Option<T> => ({
  __type__: 'some',
  value,
})
