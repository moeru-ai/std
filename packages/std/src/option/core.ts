export interface None {
  __mo: 'none'
}

export type Option<T> = None | Some<T>

export interface Some<T> {
  __mo: 'some'
  value: T
}

export const none: Option<never> = {
  __mo: 'none',
}

export const some = <T>(value: T): Option<T> => ({
  __mo: 'some',
  value,
})
