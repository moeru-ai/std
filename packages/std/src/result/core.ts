export interface Err<E> {
  __mr: 'err'
  error: E
}

export interface Ok<T> {
  __mr: 'ok'
  value: T
}

export type Result<T, E> = Err<E> | Ok<T>

export const err = <T, E>(error: E): Result<T, E> => ({
  __mr: 'err',
  error,
})

export const ok = <T, E>(value: T): Result<T, E> => ({
  __mr: 'ok',
  value,
})
