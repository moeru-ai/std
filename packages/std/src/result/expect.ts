import type { Result } from './core'

import { isOk, isErr } from './is'

export const expect = <T, E>(r: Result<T, E>, msg: string): T => {
  if (isOk(r))
    return r.value
  else
    throw new Error(msg, { cause: r.error })
}

export const expectErr = <T, E>(r: Result<T, E>, msg: string): E => {
  if (isErr(r))
    return r.error
  else
    throw new Error(msg, { cause: r.value })
}

