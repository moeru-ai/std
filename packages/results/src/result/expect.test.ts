import { describe, expect, it } from 'vitest'

import { err, ok } from '../core'
import { expectErr, expect as expectOk } from './expect'

describe('@moeru/results', () => {
  // https://doc.rust-lang.org/std/result/enum.Result.html#examples-18
  it('result.expect', () => {
    expect(() => expectOk(err('emergency failure'), 'Testing expect')).toThrowErrorMatchingSnapshot()
  })

  // https://doc.rust-lang.org/std/result/enum.Result.html#examples-21
  it('result.expectErr', () => {
    expect(() => expectErr(ok(10), 'Testing expectErr')).toThrowErrorMatchingSnapshot()
  })
})
