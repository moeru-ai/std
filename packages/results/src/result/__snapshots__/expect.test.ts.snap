// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@moeru/results > result.expect 1`] = `[Error: Testing expect]`;

exports[`@moeru/results > result.expectErr 1`] = `[Error: Testing expectErr]`;
